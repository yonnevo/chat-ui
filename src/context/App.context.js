import React, { useState, createContext, useEffect } from "react";
import avatars from "../data/avatarImgs";
import useChat from "../components/_useChat";

const AppContext = createContext();
const { Provider } = AppContext;

const AppProvider = ({ children }) => {
  const [user_details, set_details] = useState({});
  const { messages, send_message } = useChat();

  const set_name = name => {
    set_details(details => ({ ...details, name: name }));
  };

  const getRandomAvatar = () => avatars[Math.floor(Math.random() * 5)];

  useEffect(() => {
    const item = localStorage.getItem("user_details");
    if (item) {
      set_details(JSON.parse(item));
    } else {
      set_details({ name: "", avatar: getRandomAvatar() });
    }
  }, []);

  useEffect(
    () => localStorage.setItem("user_details", JSON.stringify(user_details)),
    [user_details]
  );

  // state = values to display
  const state = {
    user_details,
    messages
  };
  // actions = callbacks to invoke
  const actions = {
    set_name,
    send_message
  };

  return <Provider value={{ ...state, ...actions }}>{children}</Provider>;
};

export { AppProvider, AppContext };
