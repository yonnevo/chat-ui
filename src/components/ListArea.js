import React, { useContext, useEffect, useRef } from "react";
import { AppContext } from "../context/App.context";

import styled from "styled-components";

const ListArea = () => {
  const { messages, user_details } = useContext(AppContext);
  const { name } = user_details;

  const messagesEndRef = useRef(null)

  const scrollToBottom = () => {
    messagesEndRef.current.scrollIntoView({ behavior: "smooth" })
  }

  useEffect(scrollToBottom, [messages]);


  return (
    <Container data-testid="container">
      {messages.length
        ? messages.map((x, i) => (
            <ChatLine key={i} isMe={x.username === name}>
              <Img src={x.avatar} alt="Avatar"></Img>{" "}
              <Message>
                {x.username}: {x.text}
              </Message>
            </ChatLine>
          ))
        : "no messages"}
        <div ref={messagesEndRef} />
    </Container>
  );
};

export default ListArea;

const Container = styled.div`
  margin: 2% 10%;
  height: 60vh;
  border: 2px solid #0082dd;
  overflow: auto;
`;

const ChatLine = styled.div`
  padding: 0.2em;
  background: ${props => (props.isMe ? "#fefbd8" : "inherit")};
`;

const Img = styled.img`
  height: 40px;
  vertical-align: middle;
`;

const Message = styled.span`
  padding: 0.2em;
`;
