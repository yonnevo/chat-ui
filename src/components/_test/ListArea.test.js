import React from "react";
import ReactDOM from "react-dom";
import ListArea from "../ListArea";
import { AppProvider } from "../../context/App.context";

import renderer from 'react-test-renderer';
import "@testing-library/jest-dom/extend-expect";

it("renders without crashing", () => {
  const div = document.createElement("div");
  <AppProvider>
    ReactDOM.render(
    <ListArea />, div );
  </AppProvider>;
  ReactDOM.unmountComponentAtNode(div);
});

it("snapshot test - renders correctly", () => {
  const tree = renderer
    .create(
      <AppProvider>
        <ListArea />
      </AppProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
