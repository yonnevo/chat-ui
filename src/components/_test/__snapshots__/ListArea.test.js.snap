// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test - renders correctly 1`] = `
<div
  className="sc-bdVaJa dttLff"
  data-testid="container"
>
  no messages
</div>
`;
